### PR TITLE
Create Parser.cpp

### PR DESCRIPTION
**Pull request by will889889**
Create Parser.cpp with implementation.
Clang format Parser.h and Parser.cpp

### CODE REVIEW

File: src/Parser.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#ifndef SHORTLINK_SRC_PARSER_CPP_
+#define SHORTLINK_SRC_PARSER_CPP_
+#include "Parser.h"
+
+#include "RequestData.h"
+
+namespace shortlink {
+
+// Constructor
+Parser::Parser() {}
```
Comment (john0312) [inline src/Parser.cpp:16]: = default

File: src/Parser.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+// Destructor
+Parser::~Parser() {}
+
+// Decodes the input string, returns the RequestData.
+RequestData Parser::Parse(const std::string &rawData) {
+  std::stringstream ss = std::stringstream(rawData);
+  std::string method_str;
+  std::string target_str;
+
+  ss >> method_str;
```
Comment (john0312) [inline src/Parser.cpp:27]: Check for parse error.

File: src/Parser.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  ss >> target_str;
+
+  RequestType type = RequestType::Search;
+  if (method_str == "Create") {
+    type = RequestType::Create;
+  } else if (method_str != "Search") {
+    // error: method not valid.
+    type = RequestType::Invalid;
+  }
```
Comment (john0312) [inline src/Parser.cpp:36]: Check for error.

File: src/Parser.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+namespace shortlink {
+
+// Constructor
+Parser::Parser() {}
+
+// Destructor
+Parser::~Parser() {}
+
+// Decodes the input string, returns the RequestData.
+RequestData Parser::Parse(const std::string &rawData) {
```
Comment (john0312) [inline src/Parser.cpp:22]: unique_ptr?

**Review by john0312 — COMMENTED** [review]
Please rebase, split off clang-format changes and add unit test.

File: src/Parser.cpp (hunk 1 above)
Comment (will889889) [reply src/Parser.cpp]: Done

File: src/Parser.cpp (hunk 4 above)
Comment (will889889) [reply src/Parser.cpp]: done

File: src/Parser.cpp (hunk 2 above)
Comment (will889889) [reply src/Parser.cpp]: done
(and tested in unit test)

File: src/Parser.cpp (hunk 3 above)
Comment (will889889) [reply src/Parser.cpp:37]: done
(and tested in unit test)

**Comment by will889889** [comment]
Now this PR need to base on PR #30 , so that it can compiled.

File: CMakeLists.txt [hunk 5]
```diff
@@ -11,6 +11,40 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 set(SRC_DIR ${PROJECT_SOURCE_DIR}/src)
 
+# For google test
+include(FetchContent)
```
Comment (john0312) [inline CMakeLists.txt:15]: Either remove this or set it on WIN32 only.

File: CMakeLists.txt [hunk 6]
```diff
@@ -11,6 +11,40 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
...
+FetchContent_Declare(
+  googletest
+  URL https://github.com/google/googletest/archive/609281088cfefc76f9d0ce82e1ff6c30cc3591e5.zip
+)
+# For Windows: Prevent overriding the parent project's compiler/linker settings
+set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
```
Comment (john0312) [inline CMakeLists.txt:21]: Please remember to remove this.

File: src/Parser.h [hunk 7]
```diff
@@ -6,14 +6,14 @@ found in the LICENSE file. See the AUTHORS file for names of contributors.
...
 #include <string>
 
 namespace shortlink {
 
 // forward declare
 class RequestData;
-class SURL;
 
 // This class represents a operation that can translate raw message to
 // executable information of a request. Example:
```
Comment (john0312) [inline src/Parser.h:19]: Maybe add in the comment that the raw data is not an HTTP request, but something that is part of the HTTP request.

File: src/UnitTestParser.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+TEST(ParserTest, ParseEmpty) {
+  std::string rawData = "";
+  std::unique_ptr<shortlink::RequestData> data =
+      shortlink::Parser::Parse(rawData);
+
+  std::string target_content = "";
+  EXPECT_EQ(data->requestMethod, shortlink::RequestType::kInvalid);
+  EXPECT_EQ(data->target, target_content);
+}
```
Comment (john0312) [inline src/UnitTestParser.cpp:60]: Please add newline.

Comment (will889889) [reply src/UnitTestParser.cpp]: Added

File: src/Parser.h (hunk 7 above)
Comment (will889889) [reply src/Parser.h]: Added comments about it.

File: CMakeLists.txt (hunk 5 above)
Comment (will889889) [reply CMakeLists.txt]: Add on WIN32 only

**Review by john0312 — APPROVED** [review]

**Pull request merged**